File: B트랙/7주차/참고 코드/merge_sort.cpp
#include <bits/stdc++.h>
using namespace std;

void merge(vector<int>& ary, int left, int mid, int right)
{
	int n1 = mid - left + 1, n2 = right - mid;
	vector<int>L(n1), R(n2);
	for (int i = 0; i < n1; ++i)
	{
		L[i] = ary[left + i];
	}
	for (int i = 0; i < n2; ++i)
	{
		R[i] = ary[mid + i + 1];
	}
	int i = 0, j = 0, k = left;
	while (i < n1 && j < n2)
	{
		if (L[i] <= R[j])
		{
			ary[k++] = L[i++];
		}
		else
		{
			ary[k++] = R[j++];
		}
	}
	while (i < n1)
	{
		ary[k++] = L[i++];
	}
	while (j < n2)
	{
		ary[k++] = R[j++];
	}
}

void merge_sort(vector<int>& ary, int left, int right)
{
	if (left < right)
	{
		int mid = (left + right) / 2;
		merge_sort(ary, left, mid);
		merge_sort(ary, mid + 1, right);
		merge(ary, left, mid, right);
	}
}

int main(void)
{
	cin.tie(0)->sync_with_stdio(0);
	int n = 0;
	cin >> n;
	vector<int>ary(n);
	for (int i = 0; i < n; ++i)
	{
		cin >> ary[i];
	}
	merge_sort(ary, 0, n - 1);
	for (int i = 0; i < n; ++i)
	{
		cout << ary[i] << ' ';
	}
	return 0;
}

File: B트랙/7주차/참고 코드/merge_sort.py
import sys
input = sys.stdin.readline

def merge(left,mid,right):
    L = ary[left:mid + 1]
    R = ary[mid+1: right + 1]
    i, j, k = 0, 0, left
    while i < len(L) and j < len(R):
        if L[i] <= R[j]:
            ary[k] = L[i]
            i += 1
        else:
            ary[k] = R[j]
            j += 1
        k += 1
    while i < len(L):
        ary[k] = L[i]
        i += 1
        k += 1
    while j < len(R):
        ary[k] = R[j]
        j += 1
        k += 1

def merge_sort(left,right):
    if left < right:
        mid = (left + right) // 2
        merge_sort(left, mid)
        merge_sort(mid+1,right)
        merge(left,mid,right)

n = int(input().strip())
sys.setrecursionlimit(n+100)
ary = list(map(int,input().strip().split()))
merge_sort(0,n-1)
print(' '.join(map(str,ary)))
